### DISCUSSION TITLE
Attaching to a container and exiting/stopping on macOS

### DISCUSSION THREAD

**Discussion by jacksonbenete** [containers/podman#12906] [Q&A] [answered]
Hello there,

When I was searching about the differences between attach and exec, I understood that attaching is "recovering" the control over the process of PID 1 in a container.
It was written that one of the main differences though, is that exiting the container while "attached" would exit the process and stop the container, which makes sense, while you could exit a new started process like "bash" without affecting the rest of the container.
Then I decided to play around to see and understand the differences by myself.

I'm using the default macOS Terminal app, and the default shell zsh.

I've created a pod
`podman pod create --name db_pod -p 5432:5432`
And I created an additional container inside it
`podman run --name psql --pod db_pod -e POSTGRES_PASSWORD=postgres docker.io/library/postgres`

Everything up and running, nice.

Now, I attach
`podman attach psql`
I can confirm it works by creating another bash process in a second Terminal, and forcing syntax error on psql, I'm receiving those errors on my attached terminal session.

Now, I try a ctrl-p ctrl-q like noted in the man pages and nothing happens.
I think "I'm screwed". ;)
Ok, whatever, ctrl-c.
`podman ps`
The container is running...
`podman attach psql`
Attached again, can see logs... ctrl-d, nothing happens. Tried ctrl-\ and I received an error listing the CPU registers and stack pointer and everything, I thought "now I've managed to exit the thing".
`podman ps`
Container still running, can attach, see logs and do everything again.

I would guess that podman on macOS is receiving and parsing the keystrokes first instead of the attached container? Because... both ctrl-c and ctrl-\ should terminate the process right?

So maybe that's because we have the virtual machine layer on macOS? Or maybe it's something about the Terminal app? I just tried bash and ctrl-p, ctrl-q didn't worked as well, so I guess zsh isn't the problem.

I didn't wanted to open an issue on github because I don't know if I didn't understood attach/detach or if is something that only happens on macOS. I don't have other machine to test right now.

I posted this on reddit and someone said I could start a discussion here instead of an issue.

**Comment by mheon** [comment, marked as answer]
This is because you didn't create the container with a terminal (`-t`). Without a terminal, OS X is parsing all keystrokes, then subsequently passing them over to the container; for containers with a terminal, we set raw mode on the terminal in OS X and pass over the literal bytes from the keyboard. C-p C-q to detach does not work without `-t` because we can't interpret it ourselves, OS X is doing so.
